### PR TITLE
(closes #717) Remove unused code

### LINKED ISSUE #717: Where statement array syntax is limited
Array syntax in Fortran where statements must be for the full extent i.e. the dimension must only contain a `:`. If it is for a subset of the data in the dimension e.g. `2:4` then the PSyIR front-end raises an exception.
We also do not currently support use of 'bare' arrays, e.g. `WHERE(some_array < 0)` because we rely on array notation to identify array accesses. Once #500 is complete we will have the necessary functionality to look-up the shape of such arrays.

### PR DESCRIPTION
**Pull request by sergisiso**
#717 was the same than #2687 but there were some TODOs associated to it. Looking at them it turns out that they where related to the array shape querying ulility functionality in fparser2, but these are now unused (other than in its own tests) as we use the ArrayMixin utility methods.

### CODE REVIEW

**Comment by sergisiso** [comment]
This is ready for review. Initially I just wanted to delete some unused test, but then realised that the WHERE fix also solves the last passthrough file in NEMOv4 🥳  and it has a TODO to this issue.

After removing it from the excluded file another issue hit that file: the fact that we cannot distinguish between imported calls with arguments and arrays. This was somewhat fixed in ACCKernels by listing all the NEMO FUNCTIONS, so I generalised that into utils so it can be used by all transformations.

It passed the integration tests with expected performance.

One for @arporter (since it touches the NEMO ACC Kernels) or @LonelyCat124

File: examples/nemo/scripts/utils.py [hunk 1]
```diff
@@ -139,15 +146,18 @@ def enhance_tree_information(schedule):
...
-            reference.replace_with(call)
+        elif reference.symbol.name in NEMO_FUNCTIONS:
+            if reference.symbol.is_import or reference.symbol.is_unresolved:
+                # The parser gets these wrong, they are Calls not ArrayRefs
+                if not isinstance(reference.symbol, RoutineSymbol):
+                    # We haven't already specialised this Symbol.
+                    reference.symbol.specialise(RoutineSymbol)
+                if not (isinstance(reference.parent, Call) and
+                        reference.parent.routine is reference):
+                    call = Call.create(reference.symbol)
```
Comment (arporter) [inline examples/nemo/scripts/utils.py:157]: Comment please, e.g. "Replace the existing Reference with a Call"

File: examples/nemo/scripts/utils.py [hunk 2]
```diff
@@ -226,9 +236,10 @@ def normalise_loops(
         statements out of the loop nest.
     '''
 
-    # TODO #1902: NEMO4 mpi_ini.f90 has a HoistLocalArraysTrans bug
-    if hoist_local_arrays and schedule.root.name != "mpp_ini.f90":
-        # Apply the HoistLocalArraysTrans when possible
+    if hoist_local_arrays and schedule.name not in CONTAINS_STMT_FUNCTIONS:
+        # Apply the HoistLocalArraysTrans when possible, it cannot be applied
+        # to files with statement functions because it will attempt to put the
+        # allocate avobe it, which is not valid Fortran.
```
Comment (arporter) [inline examples/nemo/scripts/utils.py:242]: "above"

File: src/psyclone/psyir/frontend/fparser2.py [hunk 3]
```diff
@@ -396,179 +396,6 @@ def _find_or_create_psyclone_internal_cmp(node):
         "not have an ancestor container in which to add it.")
```
Comment (arporter) [inline src/psyclone/psyir/frontend/fparser2.py:398]: Nice!

File: examples/nemo/scripts/omp_gpu_trans.py [hunk 4]
```diff
@@ -98,14 +89,14 @@ def trans(psyir):
...
 
-        # In the lib_fortran file we annotate each routine of the SIGN_*
-        # interface with the OpenMP Declare Target Directive
-        if psyir.name == "lib_fortran.f90":
-            if subroutine.name.lower().startswith("sign_"):
-                OMPDeclareTargetTrans().apply(subroutine)
-                # We continue parallelising inside the routine, but this could
-                # change if the parallelisation directive are not nestable, in
-                # which case we could add a 'continue' here
+        # This are functions that are called from inside parallel regions,
```
Comment (arporter) [inline examples/nemo/scripts/omp_gpu_trans.py:92]: s/This/These/

File: examples/nemo/scripts/acc_loops_trans.py [hunk 5]
```diff
@@ -108,12 +99,12 @@ def trans(psyir):
                 hoist_expressions=True
         )
 
-        # In the lib_fortran file we annotate each routine of the SIGN_*
-        # interface with the OpenACC Routine Directive
-        if psyir.name == "lib_fortran.f90":
-            if subroutine.name.lower().startswith("sign_"):
-                ACCRoutineTrans().apply(subroutine)
-                continue
+        # This are functions that are called from inside parallel regions,
```
Comment (arporter) [inline examples/nemo/scripts/acc_loops_trans.py:102]: s/This/These/

**Review by arporter — CHANGES_REQUESTED** [review]
Nice work Sergi. There ought to be some prize for the PR which removes the most code. Tests all OK, integration tests look good. Reference guide unchanged.
Coverage unaffected.
Just a few minor things to tidy and then this can go on.

File: examples/nemo/scripts/utils.py (hunk 2 above)
Comment (sergisiso) [reply examples/nemo/scripts/utils.py:242]: Fixed

File: examples/nemo/scripts/utils.py (hunk 1 above)
Comment (sergisiso) [reply examples/nemo/scripts/utils.py:157]: Done

File: examples/nemo/scripts/omp_gpu_trans.py (hunk 4 above)
Comment (sergisiso) [reply examples/nemo/scripts/omp_gpu_trans.py:92]: Done

File: examples/nemo/scripts/acc_loops_trans.py (hunk 5 above)
Comment (sergisiso) [reply examples/nemo/scripts/acc_loops_trans.py:102]: Done

**Comment by sergisiso** [comment]
@arporter This is ready for another review

**Review by arporter — APPROVED** [review]
All good now. I've re-triggered the integration tests as master has moved on since they were last run for this branch.

**Comment by arporter** [comment]
Integration tests were green.

**Pull request merged**